### PR TITLE
feat(cors): Allowed more wildcard options

### PR DESCRIPTION
**Pull request by jaynis**
A few weeks ago the allowed CORS origins have been changed from a regex to a wildcard notation (#2389). Implementation wise all kinds of wildcards are supported, however, the validation regex on the `SecurityPolicy` CRD limits the CORS options to hostnames prefixed with an wildcard followed by a dot, allowing all subdomains of that host. This reduces the freedom when allowing cross origins a lot compared to how it was before. This PR aims to relax the validation regex a bit to enable the following use cases:

- Allowing all hosts of an specific scheme (`https://*`)
- Allowing all hosts regardless of the scheme (`*`)
- Allowing all ports of a specific host (`http://localhost:*`)

While allowing all hosts in the context of CORS might sound a bit hacky, this is sometimes required. For instance when a web service provides an API which is consumed by many third-party web applications hosted under arbitrary domains not under the control of the maintainer of aforementioned web service. In addition to that it can be very useful during application development. This is why I have added the option to allow all ports of a specific host as well.

Review the [new](https://regex101.com/r/rg05j7/2) and the [old](https://regex101.com/r/rg05j7/1) validation regexes.

### CODE REVIEW

File: api/v1alpha1/cors_types.go [hunk 1]
```diff
@@ -7,20 +7,26 @@ package v1alpha1
 
 import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+
 // Origin is defined by the scheme (protocol), hostname (domain), and port of
 // the URL used to access it. The hostname can be “precise” which is just the
 // domain name or “wildcard” which is a domain name prefixed with a single
-// wildcard label such as “*.example.com”.
+// wildcard label such as “*.example.com”. The optional port can be a wildcard
```
[the remark this answers is not in the document]
Comment (shawnh2) [reply api/v1alpha1/cors_types.go:14]: there're still some quotes that in the wrong character encoding, can you please help fixing them? thanks

[the remark this answers is not in the document]
Comment (jaynis) [reply api/v1alpha1/cors_types.go]: done

**Comment by arkodg** [comment]
@jaynis can sign your commits and repush ? DCO is failing

**Comment by zhaohuabing** [comment]
@jaynis Thanks for the improvement in the wildcard host matching. The implementation looks good to me. 

I only have a little hesitation about the port wildcard matching. Suffix/Port wildcard matching is not a common practice for hostnames. Do you have many ports for a given hostname?

**Review by zhaohuabing — COMMENTED** [review]
@jaynis Thanks for the improvement in the wildcard host matching. The implementation looks good to me.

I only have a little hesitation about the port wildcard matching. Suffix/Port wildcard matching is not a common practice for hostnames. Do you have many ports for a given hostname?

**Review by shawnh2 — APPROVED** [review]
lgtm, defer to @zhaohuabing

**Comment by jaynis** [comment]
Thank you for your review @zhaohuabing.
> I only have a little hesitation about the port wildcard matching. Suffix/Port wildcard matching is not a common practice for hostnames. In your use case, do you have many ports for a given hostname? 

The port range matching was solely meant to be a dev feature so that one can configure CORS for a host (e.g. localhost) regardless of the port the application runs on. But this scenario could be covered by the general wildcard as well, therefore I would also be fine with deleting it again if you think it is not required. Just let me know your preference.

**Comment by zhaohuabing** [comment]
> Thank you for your review @zhaohuabing.
> > I only have a little hesitation about the port wildcard matching. Suffix/Port wildcard matching is not a common practice for hostnames. In your use case, do you have many ports for a given hostname? 
> 
> The port range matching was solely meant to be a dev feature so that one can configure CORS for a host (e.g. localhost) regardless of the port the application runs on. But this scenario could be covered by the general wildcard as well, therefore I would also be fine with deleting it again if you think it is not required. Just let me know your preference.
> 
> 

Prefer to remove the suffix matching to keep it aligned with the common practice. Thanks.

**Review by zhaohuabing — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**